qt-cpp: Add dedicated test infrastructure for CMake Presets

OrkunTokdemir opened the pull request:
This patch introduces separate testing for CMake Presets integration,
ensuring the extension correctly handles CMakePresets.json configurations.

Changes include:
- Added presets.test.mts with comprehensive Presets integration test
  - Creates CMakePresets.json with qt-debug preset dynamically
  - Verifies configure, build, and SELECTED_QT_PATHS detection
  - Tests preset-based workflow with CMAKE_PREFIX_PATH
- Added runTest.presets.mts runner using prerelease CMake Tools
  - Required for CMake Presets support (stable version lacks features)
  - Uses temporary workspace to avoid polluting test environment
- Added index-presets.mts to run only presets.test.js
- Updated test helper with setCMakeConfigurationForPlatform() utility
- Excluded presets.test.js from regular test suite in index.mts
- Added npm scripts: test:presets, test:only:presets, run-test-presets

The test verifies that the extension properly detects Qt paths when using
CMake Presets instead of traditional kits-based configuration.

Task-number: [VSCODEEXT-64](https://bugreports.qt.io/browse/VSCODEEXT-64)

OrkunTokdemir committed:
qt-qml: qt-cpp: qt-lib: Add support for installing pre-release extensions in tests

Update installExtensionWithRetry to accept ExtensionInstallInfo object
with optional preRelease flag. This allows test runners to install
pre-release versions of extensions like CMake Tools when needed.
Update all test runners to use the new interface.

OrkunTokdemir committed:
qt-cpp: Add dedicated test infrastructure for CMake Presets

This patch introduces separate testing for CMake Presets integration,
ensuring the extension correctly handles CMakePresets.json configurations.

Changes include:
- Added presets.test.mts with comprehensive Presets integration test
  - Creates CMakePresets.json with qt-debug preset dynamically
  - Verifies configure, build, and SELECTED_QT_PATHS detection
  - Tests preset-based workflow with CMAKE_PREFIX_PATH
- Added runTest.presets.mts runner using prerelease CMake Tools
  - Required for CMake Presets support (stable version lacks features)
  - Uses temporary workspace to avoid polluting test environment
- Added index-presets.mts to run only presets.test.js
- Updated test helper with setCMakeConfigurationForPlatform() utility
- Excluded presets.test.js from regular test suite in index.mts
- Added npm scripts: test:presets, test:only:presets, run-test-presets

The test verifies that the extension properly detects Qt paths when using
CMake Presets instead of traditional kits-based configuration.

Task-number: [VSCODEEXT-64](https://bugreports.qt.io/browse/VSCODEEXT-64)

lugerard commented on qt-cpp/test/runTest.mts:
```diff
@@ -116,7 +117,10 @@ async function main() {
     const quietArgs = [...args, ...getQuietVSCodeArgs()];
-    const required = ['ms-vscode.cmake-tools', localQtCoreVsix];
+    const required: ExtensionInstallInfo[] = [
+      { idOrVsix: 'ms-vscode.cmake-tools', preRelease: true },
```
Is preRelease needed here? I don't see commands.test.mts and extension.test.mts depending on it.

leena-miettinen reviewed (commented):
Looks good to me.

lugerard commented on qt-cpp/test/helper.mts:
```diff
@@ -299,10 +299,24 @@ export async function cleanBuildDir(
   return buildDir;
 }
 
+export async function setCMakeConfigurationForPlatform(
```
couldn't we set up this function so settingName defaults to generator and value if value stays unknown it is set with getPlatformCMakeGenerator(). This way we would have only one setCMakeGeneratorForPlatform function.

OrkunTokdemir replied:
I added a CMakeConfigurator helper in my latest commit that tracks and resets settings automatically. I think that's a better approach than combining these functions, but I can set generator with `CMakeConfigurator` in a different commit https://github.com/qt-labs/vscodeext/commit/0a70c24f9b1fec9f26a15877e049a841f412563b

lugerard replied:
I like how class CMakeConfigrator  allows for easy reset. Will you replace both setCMakeGeneratorForPlatform functions with this class functionality?

OrkunTokdemir replied:
Done https://github.com/qt-labs/vscodeext/pull/290/commits/f8ec7be125159e578af791f1e44362aa4547c936

lugerard commented on qt-cpp/test/runTest.build.mts:
```diff
@@ -31,7 +31,11 @@ async function main() {
     });
-    installRequiredExtensions(cli, args, localQtCoreVsix);
+    const extensions = [
+      { idOrVsix: 'ms-vscode.cmake-tools' },
```
Since you are not modifying test.build.mts, I see no reason to modify runTest.buils.mts.

OrkunTokdemir replied:
I changed the `installRequiredExtensions()` implementation so that we can pass parameters dynamically. That's why I also changed it here to adapt to the new implementation.

lugerard replied:
ok!

lugerard commented on qt-cpp/test/runTest.mts:
```diff
@@ -27,7 +27,11 @@ async function main() {
     setupVSCodeSettings(userDataDir, qtRoot);
-    installRequiredExtensions(cli, args, localQtCoreVsix);
+    const extensions = [
+      { idOrVsix: 'ms-vscode.cmake-tools' },
```
I don't think we need cmake-tools for the extension test. Again since you did not change test.extension.mts I don't see why you need to change runTest.mts.

OrkunTokdemir replied:
qt-cpp depends on `ms-vscode.cmake-tools`. I cannot remove it from there. Otherwise, the test fails with the below error.
```
✔ Validated version: 1.106.3
✔ Found at https://update.code.visualstudio.com/1.106.3/linux-x64/stable?released=true
✔ Downloaded VS Code into /home/orkun/qt_work/vscodeext/qt-cpp/out/test/.vscode-test/vscode-linux-x64-1.106.3
[runTest] Wrote settings to: /home/orkun/qt_work/vscodeext/qt-cpp/out/test/.vscode-test/user-data/User/settings.json
[runTest] Missing required extensions: ms-vscode.cmake-tools

✗ qt-cpp tests failed

Error: Command failed: npm run test --  --qt-root="/home/orkun/Qt"
    at genericNodeError (node:internal/errors:984:15)
    at wrappedFn (node:internal/errors:538:14)
    at checkExecSyncError (node:child_process:891:11)
    at execSync (node:child_process:963:15)
    at runTest (/home/orkun/qt_work/vscodeext/scripts/test.ts:62:13)
    at main (/home/orkun/qt_work/vscodeext/scripts/test.ts:32:5)
```

lugerard replied:
How come this line was not present before, is it because of this other change you made in installRequiredExtensions()?

OrkunTokdemir replied:
Yes, because I changed the implementation of it.

lugerard commented on qt-cpp/test/runTest.natvis.mts:
```diff
@@ -4,17 +4,8 @@
-
 import { downloadAndUnzipVSCode, runTests } from '@vscode/test-electron';
 
-import { getQuietVSCodeArgs } from '../../qt-lib/src/test-constants.js';
```
Why changing runTest.natvis.mts since test.natvis.mts is not modified. Clean up for this file belong to a dedicated comit in my point of view.

OrkunTokdemir replied:
I had to remove this line because `getQuietVSCodeArgs()` was deleted at https://github.com/qt-labs/vscodeext/pull/290/commits/4778d1e07ad35daa0f9f29e172b047ffbfdff5c2#diff-0a4010f4483e24a7d464e99a8231847e2df602179d0453311f64ff24a384f24dL51

lugerard replied:
I see. Thanks.

lugerard commented on qt-cpp/test/runTest.presets.mts:
```diff
@@ -0,0 +1,70 @@
+// Copyright (C) 2025 The Qt Company Ltd.
+// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only
+
```
looks good to me.

lugerard commented on qt-cpp/test/suite/presets.test.mts:
```diff
@@ -0,0 +1,184 @@
+      if (fs.existsSync(presetsPath)) {
+        fs.unlinkSync(presetsPath);
+      }
+      await setCMakeConfigurationForPlatform(
```
don't you want to reset the other cmake configuration that were modified? 'configureOnOpen' and 'automaticReconfigure'

OrkunTokdemir replied:
Thanks! It is a nice catch. I added a new class to reset them automatically here https://github.com/qt-labs/vscodeext/pull/290/commits/0a70c24f9b1fec9f26a15877e049a841f412563b

lugerard replied:
This is a nice solution.

lugerard reviewed (changes requested):
Looks good overall. Tiny suggestions and some changes related to other tests that I don't think are necessary.

OrkunTokdemir committed:
qt-cpp:presets-test: Add CMakeConfigurator helper for test cleanup

The presets test was manually resetting CMake configuration settings,
which could leave the workspace in an inconsistent state if cleanup failed.

This patch introduces CMakeConfigrator helper class that:
- Tracks original values when setting CMake configuration
- Provides resetAll() method to restore all modified settings
- Ensures proper cleanup even if the test fails partway through

The presets test now uses cmakeConfigForWorkspace() to manage settings
for useCMakePresets, configureOnOpen, and automaticReconfigure,
with automatic restoration during cleanup.

Task-number: [VSCODEEXT-64](https://bugreports.qt.io/browse/VSCODEEXT-64)

OrkunTokdemir reviewed (commented):
@lugerard  Thanks for the review. I updated some parts and added new things. Can you take a look at it again?

OrkunTokdemir committed:
qt-cpp:build-test: Use CMakeConfigrator for CMake settings

The build test was using standalone helper functions to set the CMake
generator, which didn't track or reset the configuration changes.

This patch updates build.test.mts to use CMakeConfigrator:
- Replaced setCMakeGeneratorForPlatform() with cmakeConfigForWorkspace()
- Added resetAll() call in cleanup to restore original settings
- Removed setCMakeGeneratorForPlatform() from helper.mts (no longer needed)
- Removed setCMakeConfigurationForPlatform() from helper.mts (superseded by CMakeConfigrator)

Both build.test.mts and presets.test.mts now use the same CMakeConfigrator
pattern for managing CMake settings with automatic cleanup.

lugerard reviewed (approved):
Thanks
